web-lib-funcs.pl

```diff
@@ -12997,6 +12997,142 @@ sub get_default_system_locale
 return $locale_def;
 }
 
+=head2 get_http_redirect(url, [page], [timeout])
+
+Check if given URL redirects somewhere
+
+The parameters are :
+
+=item url - Given URL to check if it redirects anywhere, e.g. https://google.com
+
+=item page - Path in URL, e.g. /about as in https://www.google.com/about
+
+=item timeout - Timeout for connections, defaults to 30s
+
+=cut
+
+sub get_http_redirect
+{
+my ($url, $page, $timeout) = @_;
+my ($out, $error, $con_err);
+my ($proto, $host, $uport) = $url =~ /^(https?):\/\/([^:\/?#]*)(?:\:([0-9]+))?/;
+my ($ssl, $port) = (0, undef);
+$port = '80' if (!$uport);
+if ($proto eq 'https') {
+	$ssl = 1;
+	$port = '443' if (!$uport);
+	}
+
+# Set default port, page and timeout
+$port ||= $uport;
+$page ||= '/';
+$timeout ||= 30;
+
+# Original request
+my %rs;
+$rs{'url'} = $url;
+$rs{'proto'} = $proto;
+$rs{'host'} = $host;
+$rs{'port'} = $port;
+$rs{'page'} = $page;
+eval "use LWP::UserAgent";
+if (!$@) {
+	my $browser = new LWP::UserAgent();
+	$browser->timeout($timeout);
+	$browser->ssl_opts(verify_hostname => 0,
+                       SSL_verify_mode => 0x00);
+	$browser->max_redirect(30);
+	$browser->agent("Webmin");
+	if ($page ne '/') {
+		$url =~ s/\/$//;
+		$url .= "$page";
+		}
+	my $response = $browser->get($url);
+	if (!$response->is_success()) {
+		$rs{'error'} = $response->status_line;
+		}
+	else {
+		my $request = $response->request();
+		if ($request->as_string() =~ /\s((https?):\/\/([^:\/?#]*)(?:\:([0-9]+))?(.*?))\s/mi) {
+			my $rurl = $1;
+			my $proto = $2;
+			my $host = $3;
+			my $uport = $4;
+			my $path = $5;
+			$rurl =~ s/\/$//;
+			$rs{'redir'}->{'url'} = $rurl;
+			my $port = '80' if (!$uport);
+			if ($proto eq 'https') {
+				$port = '443' if (!$uport);
+				}
+			$port ||= $uport;
+			$rs{'redir'}->{'proto'} = $proto;
+			$rs{'redir'}->{'host'} = $host;
+			$rs{'redir'}->{'port'} = $port;
+			$rs{'redir'}->{'path'} = $path;
+			}
+		}
+	}
+else {
+	# Build headers
+	my @headers;
+	push(@headers, [ "Host", $host ]);
+	push(@headers, [ "User-agent", "Webmin" ]);
+	push(@headers, [ "Accept-language", "en" ]);
+
+	# Actually download it
+	$main::download_timed_out = undef;
+	local $SIG{ALRM} = \&download_timeout;
+	alarm($timeout);
+	my $h = &make_http_connection($host, $port, $ssl, "GET", $page, \@headers);
```

Comment:
If $h is an error message string rather than a ref, we should fail the check right here.